dd-sdk-android-core/src/test/kotlin/com/datadog/android/core/internal/data/upload/DataUploadRunnableTest.kt

```diff
@@ -692,6 +708,324 @@ internal class DataUploadRunnableTest {
 
     // endregion
 
+    // region upload quality
+
+    @Test
+    fun `M send network failure upload quality event W run { response not 202 }`(forge: Forge) {
+        // Given
+        val mockRumFeature: FeatureScope = mock()
+        whenever(mockFeatureSdkCore.getFeature(Feature.RUM_FEATURE_NAME))
+            .thenReturn(mockRumFeature)
+        testedRunnable = DataUploadRunnable(
+            featureSdkCore = mockFeatureSdkCore,
+            featureName = fakeFeatureName,
+            threadPoolExecutor = mockThreadPoolExecutor,
+            storage = mockStorage,
+            dataUploader = mockDataUploader,
+            contextProvider = mockContextProvider,
+            networkInfoProvider = mockNetworkInfoProvider,
+            systemInfoProvider = mockSystemInfoProvider,
+            uploadSchedulerStrategy = mockUploadSchedulerStrategy,
+            maxBatchesPerJob = fakeMaxBatchesPerJob,
+            internalLogger = mockInternalLogger
+        )
+        val batches = forge.aList(size = 1) { aList { getForgery<RawBatchEvent>() } }
+        val batchIds: List<BatchId> = batches.map { mock() }
+        val batchMetadata = forge.aList(size = batches.size) { aNullable { aString().toByteArray() } }
+        stubStorage(batchIds, batches, batchMetadata)
+        val mockUploadStatus: UploadStatus = mock()
+        whenever(mockUploadStatus.code).thenReturn(500)
+
+        batches.forEachIndexed { index, batch ->
+            whenever(
+                mockDataUploader.upload(
+                    fakeContext,
+                    batch,
+                    batchMetadata[index],
+                    batchIds[index]
+                )
+            ) doReturn mockUploadStatus
+        }
+
+        // When
+        testedRunnable.run()
+
+        // Then
+        argumentCaptor<UploadQualityEvent> {
+            verify(mockRumFeature, times(2)).sendEvent(
+                event = capture()
+            )
+            val firstEvent = firstValue as? UploadQualityEvent
+            val secondEvent = secondValue as? UploadQualityEvent
```

Comment:
Why do we use safe cast here? I think we can use `as` directly, because we expect these to be that type.